File: sdkwork-im-sdk-flutter/composed/lib/src/events_module.dart
import 'package:backend_sdk/src/models.dart';

import 'context.dart';
import 'types.dart';

class OpenChatEventsModule {
  final OpenChatSdkContext context;

  OpenChatEventsModule(this.context);

  Future<OpenChatSendResult> publish({
    required String type,
    String? name,
    Map<String, dynamic>? data,
    Map<String, dynamic>? metadata,
    ConversationEnvelope? conversation,
    String? toUserId,
    String? groupId,
    String? uuid,
    String? replyToId,
    String? forwardFromId,
    int? clientSeq,
    String? idempotencyKey,
    Map<String, dynamic>? extra,
    bool? needReadReceipt,
  }) {
    return context.publishEvent(
      type: type,
      name: name,
      data: data,
      metadata: metadata,
      conversation: conversation,
      toUserId: toUserId,
      groupId: groupId,
      uuid: uuid,
      replyToId: replyToId,
      forwardFromId: forwardFromId,
      clientSeq: clientSeq,
      idempotencyKey: idempotencyKey,
      extra: extra,
      needReadReceipt: needReadReceipt,
    );
  }

  Future<OpenChatSendResult> publishToUser({
    required String userId,
    required String type,
    String? name,
    Map<String, dynamic>? data,
    Map<String, dynamic>? metadata,
    String? uuid,
    String? replyToId,
    String? forwardFromId,
    int? clientSeq,
    String? idempotencyKey,
    Map<String, dynamic>? extra,
    bool? needReadReceipt,
  }) {
    return publish(
      type: type,
      name: name,
      data: data,
      metadata: metadata,
      toUserId: userId,
      uuid: uuid,
      replyToId: replyToId,
      forwardFromId: forwardFromId,
      clientSeq: clientSeq,
      idempotencyKey: idempotencyKey,
      extra: extra,
      needReadReceipt: needReadReceipt,
    );
  }

  Future<OpenChatSendResult> publishToGroup({
    required String groupId,
    required String type,
    String? name,
    Map<String, dynamic>? data,
    Map<String, dynamic>? metadata,
    String? uuid,
    String? replyToId,
    String? forwardFromId,
    int? clientSeq,
    String? idempotencyKey,
    Map<String, dynamic>? extra,
    bool? needReadReceipt,
  }) {
    return publish(
      type: type,
      name: name,
      data: data,
      metadata: metadata,
      groupId: groupId,
      uuid: uuid,
      replyToId: replyToId,
      forwardFromId: forwardFromId,
      clientSeq: clientSeq,
      idempotencyKey: idempotencyKey,
      extra: extra,
      needReadReceipt: needReadReceipt,
    );
  }

  Future<OpenChatSendResult> publishGameEvent({
    required String name,
    required String groupId,
    Map<String, dynamic>? data,
    Map<String, dynamic>? metadata,
    String? uuid,
    String? replyToId,
    String? forwardFromId,
    int? clientSeq,
    String? idempotencyKey,
    Map<String, dynamic>? extra,
    bool? needReadReceipt,
  }) {
    return publish(
      type: 'GAME_EVENT',
      name: name,
      data: data,
      metadata: <String, dynamic>{
        'namespace': 'game',
        'version': 1,
        ...?metadata,
      },
      groupId: groupId,
      uuid: uuid,
      replyToId: replyToId,
      forwardFromId: forwardFromId,
      clientSeq: clientSeq,
      idempotencyKey: idempotencyKey,
      extra: extra,
      needReadReceipt: needReadReceipt,
    );
  }
}

File: sdkwork-im-sdk-typescript/composed/src/events-module.ts
import type {
  OpenChatPublishEventOptions,
} from './types';
import type { OpenChatSdkContext } from './sdk-context';

export function createEventsModule(context: OpenChatSdkContext) {
  return {
    publish(payload: OpenChatPublishEventOptions) {
      return context.publishEvent(payload);
    },

    publishToUser(payload: {
      userId: string;
      type: string;
      name?: string;
      data?: Record<string, unknown>;
      metadata?: Record<string, unknown>;
    }) {
      return context.publishEvent({
        ...payload,
        toUserId: payload.userId,
      });
    },

    publishToGroup(payload: {
      groupId: string;
      type: string;
      name?: string;
      data?: Record<string, unknown>;
      metadata?: Record<string, unknown>;
    }) {
      return context.publishEvent({
        ...payload,
        groupId: payload.groupId,
      });
    },

    publishGameEvent(payload: {
      groupId: string;
      name: string;
      data?: Record<string, unknown>;
      metadata?: Record<string, unknown>;
      uuid?: string;
      replyToId?: string;
      forwardFromId?: string;
      clientSeq?: number;
      idempotencyKey?: string;
      extra?: Record<string, unknown>;
      needReadReceipt?: boolean;
    }) {
      return context.publishEvent({
        type: 'GAME_EVENT',
        name: payload.name,
        data: payload.data,
        metadata: {
          namespace: 'game',
          version: 1,
          ...payload.metadata,
        },
        groupId: payload.groupId,
        uuid: payload.uuid,
        replyToId: payload.replyToId,
        forwardFromId: payload.forwardFromId,
        clientSeq: payload.clientSeq,
        idempotencyKey: payload.idempotencyKey,
        extra: payload.extra,
        needReadReceipt: payload.needReadReceipt,
      });
    },
  };
}
